src/goto-harness/recursive_initialization.cpp

```diff
@@ -121,7 +121,77 @@ void recursive_initializationt::initialize(
   else
   {
     initialize_nondet(lhs, depth, known_tags, body);
+
+const irep_idt &recursive_initializationt::build_constructor(const exprt &expr)
+{
+  // for `expr` of type T builds a declaration of a function:
+  //
+  // void type_constructor_T(int depth_T, T *result);
+  //
+  // or in case a `size` is associated with the `expr`
+  //
+  // void type_constructor_T(int depth_T, T *result_T, int *size);
+  optionalt<irep_idt> size_var;
+  optionalt<irep_idt> expr_name;
+  if(expr.id() == ID_symbol)
+  {
+    expr_name = to_symbol_expr(expr).get_identifier();
+    if(should_be_treated_as_array(*expr_name))
+      size_var = get_associated_size_variable(*expr_name);
+  }
+  const typet &type = expr.type();
+  if(type_constructor_names.find(type) != type_constructor_names.end())
+    return type_constructor_names[type];
+
+  const std::string &pretty_type = type2id(type);
+  symbolt &depth_symbol =
+    get_fresh_param_symbol("depth_" + pretty_type, signed_int_type());
+  depth_symbol.value = from_integer(0, signed_int_type());
+
+  code_typet::parameterst fun_params;
+  code_typet::parametert depth_parameter{signed_int_type()};
+  depth_parameter.set_identifier(depth_symbol.name);
+  fun_params.push_back(depth_parameter);
+
+  const typet &result_type = pointer_type(type);
+  const symbolt &result_symbol =
+    get_fresh_param_symbol("result_" + pretty_type, result_type);
+  code_typet::parametert result_parameter{result_type};
+  result_parameter.set_identifier(result_symbol.name);
+  fun_params.push_back(result_parameter);
+
+  optionalt<exprt> size_symbol_expr;
+  auto &symbol_table = goto_model.symbol_table;
+  if(size_var.has_value())
+  {
+    const symbol_exprt &size_var_symbol_expr =
+      symbol_table.lookup_ref(*size_var).symbol_expr();
+    const typet &size_var_type = pointer_type(size_var_symbol_expr.type());
+    const symbolt &size_symbol =
+      get_fresh_param_symbol("size_" + pretty_type, size_var_type);
+
+    fun_params.push_back(code_typet::parametert{size_var_type});
+    fun_params.back().set_identifier(size_symbol.name);
+    size_symbol_expr = size_symbol.symbol_expr();
   }
+
+  const symbolt &function_symbol = get_fresh_fun_symbol(
+    "type_constructor_" + pretty_type, code_typet{fun_params, empty_typet{}});
+  type_constructor_names[type] = function_symbol.name;
+  symbolt *mutable_symbol = symbol_table.get_writeable(function_symbol.name);
+
+  // the body is specific for each type of expression
+  mutable_symbol->value = build_constructor_body(
```

Comment:
slightly weird that `build_constructor_body` is being referenced here but added in the next commit and there's a prototype added for it as well in `recursive_initialization.h`

Reply:
I wasn't sure how to break the PR into individual commits so I sometimes formed a commit around a body of a function but only declared the functions called inside that body.